### ISSUE TITLE
Should we use Enums for method behaviors?

### ISSUE THREAD

**Issue by iwburns** [iwburns/id-tree#18]
So the idea here is to reduce the number of public methods by taking ones that are similar and condensing them down into one method that takes all of the original arguments plus an Enum for the behavior that the user wants.

For example, we currently have three different `remove_node_[something]_children` methods.  They could instead just be called `remove_node` and we could add an extra argument for specifying the desired behavior.
The Enum would be something like:
```
pub enum RemoveBehavior {
    DropChildren,
    LiftChildren,
    OrphanChildren,
}
```

We could then use this to combine any potential other behaviors of `swap_sub_tree`.  That method would be called `swap_nodes` and we could add Enum variants later for if children are left alone, if they are carried along, if they are dropped (don't know why someone would want this, but maybe they will).

I can't currently think of any downsides to this, but that doesn't mean there aren't any.
@Drakulix any thoughts here?

EDIT:  Adding more requirements to this one:

In addition to `RemoveBehavior` above, we are now going to have `MoveBehavior` and `InsertBehavior`.

**Comment by Drakulix** [comment]
I like this as long as it is only used for simple decisions (like how to handle the remaining children in case of the `remove_node` function).
This should never be used for lets say, different behaviors we discussed for `move_node_to_parent`. These method behaviors are already quite complex and the API should not become more confusing.

**Comment by iwburns** [comment]
I agree completely.  I think those behaviors should be documented and set in stone for sure.

Do you foresee any issues where adding an option to one of those Enums could break a library user's code?  I feel like it's super unlikely that anyone would ever be matching on an Enum of that nature, but I guess it's technically possible to do.

Just thinking of what constitutes a breaking change in the future if we go that route.

**Comment by Drakulix** [comment]
Either just increase the version of your crate if you do that (I mean thats what semver is for),
or add a note to the `enum`s documentation, that you either should not match that struct *or*, that this struct might be extended without increasing the semver version and a match should always include a `_ =>` pattern handling an unknown operation as appropriate.

But I would just go with increasing the version every time. I mean if you add a behavior this is basically adding a new function / new functionality, you should not try to work around semver in that case.

**Comment by iwburns** [comment]
Yeah, that's a good point.  And I don't think that adding behaviors is going to skyrocket this lib into v25.x.x anytime soon, because there are really only so many behaviors that we could add in the first place.

**Comment by iwburns** [comment]
Alright, I'm going to move forward with this.  Hopefully I'll have a (WiP) PR up by EoD today, but if not I should easily have something up by tomorrow.

**Comment by iwburns** [comment]
PR is up: #19 
Very much a WiP though

**Comment by iwburns** [comment]
Yes, I would like to go ahead and refactor those as well.

In addition, I had a thought earlier that's related to this and I wanted to get your opinion on it:
What if we combined `insert_with_parent` and `set_root` since they are both technically "insert" methods with different destinations?  We could use something like this for the enum:
```
pub enum InsertLocation {
    Root,
    UnderParent(&NodeId),  //not sure about the name on this one.
}
```

**Comment by Drakulix** [comment]
I would even suggest calling that `LocationBehavior` and use it for both `insert` and `move` methods.

**Comment by iwburns** [comment]
Oh, good idea! 😄 I'll update #28 so that the enum is usable for both `insert` and `move` and then do the `insert` changes in another PR.

**Comment by iwburns** [comment]
@Drakulix I'm wondering your opinion on a change I'm thinking of making:

Currently we have these behavior enums in the `behaviors` module:
```
pub enum RemoveBehavior {
    DropChildren,
    LiftChildren,
    OrphanChildren,
}
pub enum MoveBehavior<'a> {
    ToRoot,
    ToParent(&'a NodeId),
}
pub enum InsertBehavior<'a> {
    AsRoot,
    UnderNode(&'a NodeId),
}
```
and with the way we're `pub use`ing them individually, externally there isn't any `behaviors` module.  That means you can use them directly like this:
```
use id_tree::InsertBehavior;
let id = tree.insert(Node::new(10), InsertBehavior::UnderNode(&some_id)).ok().unwrap();
```
But I feel like that's overly wordy and results in longer lines of code.

I'm thinking about using exposing the `behaviors` module, and then removing the "Behavior" suffix from each of the behaviors.  They would then look like this:
```
pub enum Remove {
    DropChildren,
    LiftChildren,
    OrphanChildren,
}
pub enum Move<'a> {
    ToRoot,
    ToParent(&'a NodeId),
}
pub enum Insert<'a> {
    AsRoot,
    UnderNode(&'a NodeId),
}
```
but they would be scoped under `behaviors` so you could use them like this if you want:
```
use id_tree::behaviors::Insert;
let id = tree.insert(Node::new(10), Insert::UnderNode(&some_id)).ok().unwrap();
```

I'm not sure if this is worth doing or not, but it seems like it's worth discussing.  Basically I'm looking for ways to make the API a little bit more terse.  I don't want it to be overly terse because it might get too confusing, but I think it's a bit wordy as it is currently.

**Comment by iwburns** [comment]
@Drakulix After a bit more thinking on this, I realized that the caller can change the `use` statement to save space if it ends up being an issue:
```
use id_tree::*;
use id_tree::InsertBehavior::*;

//...

let id = tree.insert(Node::new(10), InsertBehavior::UnderNode(&some_id)).ok().unwrap();
//the above could then become this:
let id = tree.insert(Node::new(10), UnderNode(&some_id)).ok().unwrap();
```
I'm not sure if that's idiomatic Rust or not, but it does make me feel less worried about the verbosity that is introduced by the behavior-system that we've added.  Thoughts?

**Comment by Drakulix** [comment]
It is idomatic rust, but mostly used inside of functions rather then for the whole module.
The reasoning is, that `enum`-variants can be much more generic, then the enum itself and this might lead to confusing code.

So you usually do:
```
use id_tree::*;

fn do_much_tree_stuff() {
    use id_tree::InsertBehavior::*; //Make following code shorter and easier to grasp in this small context
    ...
    let id = tree.insert(Node::new(10), UnderNode(&some_id)).ok().unwrap();
    ...
}
...
```

For small files it is also common to import the variants directly for the whole module.
So yes, you can expect users to do these imports, you may also use them in examples and I would encourage that. This should address verbosity quite nicely.
